bin/varnishd/cache/cache_vrt_vcl.c

```diff
@@ -88,6 +88,12 @@ VCL_Refresh(struct vcl **vcc)
 	VCL_Update(vcc, NULL);
 }
 
+static inline int
+vcl_cancache(const struct vcl *vcl)
+{
+	return (vcl == vcl_active || vcl->nlabels > 0);
+}
+
```

Comment:
For lack of a better place, I will comment here about the general idea behind this change.

In a setup where the main VCL is only an empty shell that only dispatches requests to other VCLs via labels, caching anything but the active VCL is bound to be undone... on the client side. So it's not just opportunistic, it's even more of a gamble since recaching anything but the active VCL actively (pun intended) jeopardizes future client tasks, and the more the labels the lower the probability to benefit from recaching on the backend side.

Also, after 60s of inactivity an idle thread will release its VCL reference. Should we make the worker keep it instead when it's still the active VCL to trade vcl_mtx usage with the pool's mutex instead? I mean, 60s is probably less frequent than starting and completing tasks on a busy system.

Reply:
I do not know enough about the circumstances which lead to the initial vcl caching iteration (maybe @bsdphk can tell us?), but I am well aware of the situation with labels and wanted to make incremental progress.
Regarding _jeopardizes future client tasks_ I am not quite sure what you mean. Do you think keeping a reference which is likely to be useless does any harm to future requests? I would think that, since we return the old reference and get the new one in one go, the difference would be marginal, but I agree that with labels, caching only benefits the backend side and the chances for it to be beneficial are inversely proportional to the number of labels.
Regarding the idle timeout, we could make it a parameter. Other than that I am also not quite sure where you are heading with that thought.
The other bit of context is that I have a long-term TODO List item: Backend Reference Counting. A previous iteration was rightfully criticized for introducing another central lock, so my intention is to propose re-using `vcl_mtx`.
So if we needed to improve vcl caching even further to optimize the many-labels-case, I would be interested to know in order to align the other plans.